File: sanitizer/pdf_sanitizer.py
"""
PDF Sanitizer - Layer 3 (Hex-Decoded Keyword Engine)

Detection techniques borrowed from:
- PDFiD (Didier Stevens): Hex de-obfuscation (#XX decoding), keyword counting
  for /JavaScript, /JS, /OpenAction, /AA, /Launch, /EmbeddedFile, /XFA
- Dangerzone (paranoid mode concept): full object-tree rebuild keeping only
  safe rendering keys (/Page, /Pages, /Contents, /Resources, /MediaBox, etc.)

Extended with:
- /SubmitForm, /ImportData detection
- Recursive object tree traversal (including Annotations)
- Action /S sub-type analysis (/S = JavaScript | Launch | URI | SubmitForm)
"""

import os
import re
from typing import Dict, Any, Optional, Callable, Set, Union
from pypdf import PdfReader, PdfWriter
from pypdf.generic import (
    DictionaryObject, NameObject, ArrayObject,
    IndirectObject, NumberObject,
)
from .base import BaseSanitizer, Threat, SanitizeReport, SanitizeMode


class PDFSanitizer(BaseSanitizer):
    """Scans and sanitizes PDF files using Layer 3 keyword engine."""

    # ── Dangerous PDF keys (normalized names) ─────────────────────────
    # High severity: code execution / auto-action
    HIGH_KEYS = frozenset({
        "/JS", "/JavaScript", "/OpenAction", "/AA", "/Launch",
    })
    # Medium severity: data exfiltration / embedded payloads
    MEDIUM_KEYS = frozenset({
        "/EmbeddedFiles", "/EmbeddedFile", "/XFA", "/SubmitForm", "/ImportData",
        "/RichMedia", "/Rendition",
    })
    ALL_DANGEROUS_KEYS = HIGH_KEYS | MEDIUM_KEYS

    # Action sub-types that execute code or launch local programs.
    HIGH_ACTION_TYPES = frozenset({
        "/JavaScript", "/JS", "/Launch",
    })

    # Action sub-types that can open external content or exfiltrate form data.
    MEDIUM_ACTION_TYPES = frozenset({
        "/URI", "/SubmitForm", "/ImportData", "/GoToR", "/GoToE", "/Rendition",
    })

    DANGEROUS_ACTION_TYPES = HIGH_ACTION_TYPES | MEDIUM_ACTION_TYPES

    # Keys safe to keep in PARANOID mode page rebuild
    PARANOID_SAFE_PAGE_KEYS = frozenset({
        "/Type", "/Parent", "/Contents", "/Resources", "/MediaBox",
        "/CropBox", "/BleedBox", "/TrimBox", "/ArtBox", "/Rotate",
        "/UserUnit",
    })

    # Keys safe to keep in PARANOID mode root rebuild
    PARANOID_SAFE_ROOT_KEYS = frozenset({
        "/Type", "/Pages", "/PageLayout", "/PageMode",
        "/Metadata", "/MarkInfo", "/StructTreeRoot", "/Lang",
    })

    def __init__(self, file_path: str, log_callback: Optional[Callable[[str], None]] = None):
        super().__init__(file_path, log_callback)

    # ══════════════════════════════════════════════════════════════════
    #  HEX DE-OBFUSCATION (borrowed from PDFiD)
    # ══════════════════════════════════════════════════════════════════

    @staticmethod
    def _normalize_name(name: str) -> str:
        """Decode PDF hex-obfuscated names: /J#61vaScript → /JavaScript"""
        if '#' not in name:
            return name
        return re.sub(
            r'#([0-9a-fA-F]{2})',
            lambda m: chr(int(m.group(1), 16)),
            name
        )

    # ══════════════════════════════════════════════════════════════════
    #  SCAN
    # ══════════════════════════════════════════════════════════════════

    def scan(self) -> SanitizeReport:
        self.report.log(f"Starting PDF scan: {os.path.basename(self.file_path)}")
        try:
            reader = PdfReader(self.file_path)
            visited: Set[int] = set()

            # 1. Document Root (Catalog)
            if reader.root_object:
                self._scan_dict("Document Root", reader.root_object, visited)

            # 2. All pages + annotations
            for i, page in enumerate(reader.pages):
                page_obj = page.get_object()
                self._scan_dict(f"Page {i + 1}", page_obj, visited)

                annots = page_obj.get("/Annots")
                if isinstance(annots, ArrayObject):
                    for j, ref in enumerate(annots):
                        annot = ref.get_object() if hasattr(ref, 'get_object') else ref
                        if isinstance(annot, DictionaryObject):
                            self._scan_dict(f"Page {i + 1} Annot {j + 1}", annot, visited)

            self.report.success = True

        except Exception as e:
            self.report.error(f"Failed to scan PDF: {e}")

        self.report.log(
            f"Scan complete. Found {len(self.report.threats)} threat(s): "
            f"{self.report.threat_summary()}"
        )
        return self.report

    def _scan_dict(self, location: str, obj: DictionaryObject, visited: Set[int]):
        obj_id = id(obj)
        if obj_id in visited:
            return
        visited.add(obj_id)

        for key, value in obj.items():
            norm = self._normalize_name(key)

            # Check if the key itself is dangerous
            if norm in self.ALL_DANGEROUS_KEYS:
                severity = "High" if norm in self.HIGH_KEYS else "Medium"
                self.report.add_threat(Threat(
                    norm.replace("/", ""), location,
                    f"Key '{key}' (normalized: '{norm}')",
                    severity
                ))

            # Check Action sub-type: /A { /S /JavaScript ... }
            if norm == "/A":
                self._scan_action(location, value)

            # Recurse
            resolved = value.get_object() if hasattr(value, 'get_object') else value
            if isinstance(resolved, DictionaryObject):
                self._scan_dict(location, resolved, visited)
            elif isinstance(resolved, ArrayObject):
                self._scan_array(location, resolved, visited)

    def _scan_action(self, location: str, value):
        action = value.get_object() if hasattr(value, 'get_object') else value
        if not isinstance(action, DictionaryObject):
            return
        s_type = self._normalize_name(str(action.get("/S", "")))
        if s_type in self.DANGEROUS_ACTION_TYPES:
            severity = "High" if s_type in self.HIGH_ACTION_TYPES else "Medium"
            self.report.add_threat(Threat(
                "Action", location,
                f"Action /S = {s_type}",
                severity
            ))

    def _scan_array(self, location: str, arr: ArrayObject, visited: Set[int]):
        for item in arr:
            resolved = item.get_object() if hasattr(item, 'get_object') else item
            if isinstance(resolved, DictionaryObject):
                self._scan_dict(location, resolved, visited)
            elif isinstance(resolved, ArrayObject):
                self._scan_array(location, resolved, visited)

    # ══════════════════════════════════════════════════════════════════
    #  SANITIZE (three-tier modes)
    # ══════════════════════════════════════════════════════════════════

    def sanitize(self, output_path: str, mode: SanitizeMode = SanitizeMode.STANDARD, scrub_metadata: bool = False) -> SanitizeReport:
        self.report.log(f"Sanitizing PDF [{mode.value}]: {os.path.basename(self.file_path)}")

        try:
            reader = PdfReader(self.file_path)
            writer = PdfWriter()

            # ── In-place clean pages and add to writer ──
            for i, page in enumerate(reader.pages):
                if mode == SanitizeMode.PARANOID:
                    self._paranoid_rebuild_page_inplace(i, page)
                else:
                    self._clean_page_inplace(i, page, mode)

                writer.add_page(page)

            # ── Clean document root ──
            self._clean_root(reader, writer, mode, scrub_metadata)

            # ── Document Information (/Info) ──
            if scrub_metadata:
                writer._info.clear()
                self.report.log("Trailer: Cleared Document Info dictionary (/Info) for metadata scrubbing.")
            elif reader.metadata:
                writer.add_metadata(reader.metadata)

            # ── Write output ──
            with open(output_path, "wb") as f:
                writer.write(f)

            self.report.sanitized_path = output_path
            self.report.success = True
            self.report.log("PDF sanitization completed successfully.")

        except Exception as e:
            self.report.error(f"Sanitization failed: {e}")
            self.report.success = False

        return self.report

    # ── Page-level cleaning (in-place) ────────────────────────────────

    def _clean_page_inplace(self, page_idx: int, page: DictionaryObject,
                            mode: SanitizeMode):
        """Clean a page dict in-place by removing dangerous keys."""
        for k in list(page.keys()):
            norm = self._normalize_name(k)

            # All modes: strip /AA (Additional Actions)
            if norm == "/AA":
                del page[k]
                self.report.log(f"Page {page_idx + 1}: Stripped /AA")
                continue

            # Handle annotations
            if norm == "/Annots":
                annots_val = page[k]
                annots = annots_val.get_object() if hasattr(annots_val, 'get_object') else annots_val
                if isinstance(annots, ArrayObject):
                    self._clean_annotations_inplace(page_idx, annots, mode)

    def _clean_annotations_inplace(self, page_idx: int, annots: ArrayObject,
                                   mode: SanitizeMode):
        """Clean page annotations in-place."""
        for i, ref in enumerate(list(annots)):
            annot = ref.get_object() if hasattr(ref, 'get_object') else ref
            if isinstance(annot, DictionaryObject):
                self._clean_dict_obj_inplace(
                    f"Page {page_idx + 1} Annot {i + 1}", annot, mode
                )

    def _clean_dict_obj_inplace(self, location: str, obj: DictionaryObject,
                                mode: SanitizeMode):
        """Recursively clean a dictionary object in-place."""
        for k in list(obj.keys()):
            norm = self._normalize_name(k)

            # STANDARD: remove JS, AA, OpenAction, Launch
            if norm in self.HIGH_KEYS:
                del obj[k]
                self.report.log(f"{location}: Stripped {norm}")
                continue

            # STRICT / PARANOID: also remove medium-severity keys
            if mode in (SanitizeMode.STRICT, SanitizeMode.PARANOID):
                if norm in self.MEDIUM_KEYS:
                    del obj[k]
                    self.report.log(f"{location}: Stripped {norm}")
                    continue

            # Handle /A (Action) sub-dictionaries
            if norm == "/A":
                action = obj[k].get_object() if hasattr(obj[k], 'get_object') else obj[k]
                if isinstance(action, DictionaryObject):
                    s_type = self._normalize_name(str(action.get("/S", "")))

                    # Always remove code execution / local launch actions.
                    if s_type in self.HIGH_ACTION_TYPES:
                        del obj[k]
                        self.report.log(f"{location}: Stripped action /S={s_type}")
                        continue

                    # STRICT+: also remove external navigation and form/data actions.
                    if mode in (SanitizeMode.STRICT, SanitizeMode.PARANOID):
                        if s_type in self.MEDIUM_ACTION_TYPES:
                            del obj[k]
                            self.report.log(f"{location}: Stripped action /S={s_type}")
                            continue

            # Recurse into sub-dicts
            v = obj[k]
            resolved = v.get_object() if hasattr(v, 'get_object') else v
            if isinstance(resolved, DictionaryObject):
                self._clean_dict_obj_inplace(location, resolved, mode)

    # ── PARANOID: full page rebuild (in-place) ────────────────────────

    def _paranoid_rebuild_page_inplace(self, page_idx: int, page: DictionaryObject):
        """Rebuild page dict in-place keeping ONLY safe rendering keys."""
        for k in list(page.keys()):
            norm = self._normalize_name(k)
            if norm not in self.PARANOID_SAFE_PAGE_KEYS:
                del page[k]
                self.report.log(f"Page {page_idx + 1} (paranoid): Dropped {norm}")

    # ── Root (Catalog) cleaning ───────────────────────────────────────

    def _clean_root(self, reader: PdfReader, writer: PdfWriter,
                    mode: SanitizeMode, scrub_metadata: bool = False):
        """Clean the document catalog / root object."""
        root = writer._root_object
        orig_root = reader.root_object

        for k, v in orig_root.items():
            norm = self._normalize_name(k)

            # Skip keys already managed by PdfWriter
            if k in root:
                continue

            # Skip metadata stream if scrubbing
            if norm == "/Metadata" and scrub_metadata:
                self.report.log("Root: Stripped /Metadata stream (Metadata Scrubbing)")
                continue

            # STANDARD: strip high-severity root keys
            if norm in self.HIGH_KEYS:
                self.report.log(f"Root: Stripped {norm}")
                continue

            # All modes: clean /Names subtree
            if norm == "/Names":
                names = v.get_object() if hasattr(v, 'get_object') else v
                if isinstance(names, DictionaryObject):
                    clean_names = self._clean_names_tree(names, mode)
                    root[NameObject(k)] = clean_names
                    continue

            # STRICT+: strip medium-severity root keys
            if mode in (SanitizeMode.STRICT, SanitizeMode.PARANOID):
                if norm in self.MEDIUM_KEYS:
                    self.report.log(f"Root: Stripped {norm}")
                    continue

            # PARANOID: only keep safe root keys
            if mode == SanitizeMode.PARANOID:
                if norm not in self.PARANOID_SAFE_ROOT_KEYS:
                    self.report.log(f"Root (paranoid): Dropped {norm}")
                    continue

            root[NameObject(k)] = v

    def _clean_names_tree(self, names: DictionaryObject,
                          mode: SanitizeMode) -> DictionaryObject:
        """Clean the /Names dictionary (JS name tree, embedded files, etc.) in-place."""
        clean = DictionaryObject()
        for k, v in names.items():
            norm = self._normalize_name(k)

            # Always remove JS name tree
            if norm == "/JavaScript":
                self.report.log("Root/Names: Stripped /JavaScript name tree")
                continue

            # STRICT+: remove embedded files
            if mode in (SanitizeMode.STRICT, SanitizeMode.PARANOID):
                if norm == "/EmbeddedFiles":
                    self.report.log("Root/Names: Stripped /EmbeddedFiles")
                    continue

            clean[NameObject(k)] = v
        return clean

File: web/pdf-sanitizer.js
/**
 * PDF Sanitizer - Layer 3 (Hex-Decoded Keyword Engine)
 * JavaScript Port using pdf-lib for Client-side Browser execution.
 */

class PDFSanitizer {
  static HIGH_KEYS = new Set(["/JS", "/JavaScript", "/OpenAction", "/AA", "/Launch"]);
  static MEDIUM_KEYS = new Set(["/EmbeddedFiles", "/XFA", "/SubmitForm", "/ImportData"]);
  static ALL_DANGEROUS_KEYS = new Set([...PDFSanitizer.HIGH_KEYS, ...PDFSanitizer.MEDIUM_KEYS]);
  static DANGEROUS_ACTION_TYPES = new Set(["/JavaScript", "/Launch", "/SubmitForm", "/ImportData"]);

  static PARANOID_SAFE_PAGE_KEYS = new Set([
    "/Type", "/Parent", "/Contents", "/Resources", "/MediaBox",
    "/CropBox", "/BleedBox", "/TrimBox", "/ArtBox", "/Rotate",
    "/UserUnit"
  ]);

  static PARANOID_SAFE_ROOT_KEYS = new Set([
    "/Type", "/Pages", "/PageLayout", "/PageMode",
    "/Metadata", "/MarkInfo", "/StructTreeRoot", "/Lang"
  ]);

  constructor(file, logCallback = null) {
    this.file = file;
    this.logCallback = logCallback || window.currentTestLogger;
    this.threats = [];
    this.errors = [];
    this.logs = [];
    this.success = false;
  }

  log(msg) {
    this.logs.push(msg);
    if (this.logCallback) {
      this.logCallback(msg, 'info');
    }
  }

  error(msg) {
    this.errors.push(msg);
    this.logs.push(`ERROR: ${msg}`);
    if (this.logCallback) {
      this.logCallback(`ERROR: ${msg}`, 'error');
    }
  }

  addThreat(type, path, description, severity = 'High') {
    const threat = { type, path, description, severity };
    this.threats.push(threat);
    if (this.logCallback) {
      this.logCallback(`[${severity}] Found ${type} in ${path}: ${description}`, 'threat');
    }
  }

  getThreatSummary() {
    const summary = { High: 0, Medium: 0, Low: 0 };
    for (const t of this.threats) {
      if (summary[t.severity] !== undefined) {
        summary[t.severity]++;
      }
    }
    return summary;
  }

  /**
   * Decode PDF hex-obfuscated names: /J#61vaScript -> /JavaScript
   */
  _normalizeName(nameStr) {
    if (!nameStr.includes('#')) return nameStr;
    return nameStr.replace(/#([0-9a-fA-F]{2})/g, (match, hex) => {
      return String.fromCharCode(parseInt(hex, 16));
    });
  }

  _resolve(val, context) {
    if (val && typeof val.ref === 'object') { // pdf-lib PDFRef instance check
      return context.lookup(val);
    }
    return val;
  }

  /**
   * Scan PDF for threats.
   */
  async scan() {
    this.log(`Starting PDF scan: ${this.file.name}`);
    this.threats = [];
    this.errors = [];

    try {
      const arrayBuffer = await this.file.arrayBuffer();
      const pdfDoc = await PDFLib.PDFDocument.load(arrayBuffer);
      const context = pdfDoc.context;
      const visited = new Set();

      // 1. Document Root (Catalog)
      const catalog = pdfDoc.catalog;
      if (catalog) {
        this._scanDict("Document Root", catalog, context, visited);
      }

      // 2. All Pages + Annotations
      const pages = pdfDoc.getPages();
      for (let i = 0; i < pages.length; i++) {
        const page = pages[i];
        const pageObj = page.node; // Page object is a DictionaryObject
        this._scanDict(`Page ${i + 1}`, pageObj, context, visited);

        // Scan annotations array
        const annots = this._resolve(pageObj.get(PDFLib.PDFName.of('Annots')), context);
        if (annots instanceof PDFLib.PDFArray) {
          for (let j = 0; j < annots.size(); j++) {
            const annot = this._resolve(annots.get(j), context);
            if (annot instanceof PDFLib.PDFDict) {
              this._scanDict(`Page ${i + 1} Annot ${j + 1}`, annot, context, visited);
            }
          }
        }
      }

      this.log(`Scan complete. Found ${this.threats.length} threat(s).`);
      return {
        success: true,
        threats: this.threats,
        summary: this.getThreatSummary()
      };
    } catch (e) {
      this.error(`Failed to scan PDF: ${e.message}`);
      return { success: false, threats: [], summary: { High: 0, Medium: 0, Low: 0 } };
    }
  }

  _scanDict(location, obj, context, visited) {
    if (!obj || visited.has(obj)) return;
    visited.add(obj);

    const keys = obj.keys();
    for (const key of keys) {
      const norm = this._normalizeName(key.toString()); // toString yields '/Key'

      // Check if key itself is dangerous
      if (PDFSanitizer.ALL_DANGEROUS_KEYS.has(norm)) {
        const severity = PDFSanitizer.HIGH_KEYS.has(norm) ? "High" : "Medium";
        this.addThreat(
          norm.replace("/", ""),
          location,
          `Key '${key.toString()}' (normalized: '${norm}')`,
          severity
        );
      }

      // Check Action sub-type: /A { /S /JavaScript ... }
      if (norm === "/A") {
        const val = obj.get(key);
        this._scanAction(location, val, context);
      }

      // Recurse into sub-dictionaries and arrays
      const value = this._resolve(obj.get(key), context);
      if (value instanceof PDFLib.PDFDict) {
        this._scanDict(location, value, context, visited);
      } else if (value instanceof PDFLib.PDFArray) {
        this._scanArray(location, value, context, visited);
      }
    }
  }

  _scanAction(location, value, context) {
    const action = this._resolve(value, context);
    if (!(action instanceof PDFLib.PDFDict)) return;

    const sVal = this._resolve(action.get(PDFLib.PDFName.of('S')), context);
    if (sVal instanceof PDFLib.PDFName) {
      const sType = this._normalizeName(sVal.toString());
      if (PDFSanitizer.DANGEROUS_ACTION_TYPES.has(sType)) {
        this.addThreat(
          "Action",
          location,
          `Action /S = ${sType}`,
          "High"
        );
      }
    }
  }

  _scanArray(location, arr, context, visited) {
    for (let i = 0; i < arr.size(); i++) {
      const value = this._resolve(arr.get(i), context);
      if (value instanceof PDFLib.PDFDict) {
        this._scanDict(location, value, context, visited);
      } else if (value instanceof PDFLib.PDFArray) {
        this._scanArray(location, value, context, visited);
      }
    }
  }

  /**
   * Sanitize PDF and generate a clean PDF Blob.
   */
  async sanitize(mode = 'standard', scrubMetadata = false) {
    this.log(`Sanitizing PDF [${mode}]: ${this.file.name}`);
    this.threats = [];
    this.errors = [];

    try {
      const arrayBuffer = await this.file.arrayBuffer();
      const pdfDoc = await PDFLib.PDFDocument.load(arrayBuffer);
      const context = pdfDoc.context;

      // ── Clean all pages ──
      const pages = pdfDoc.getPages();
      for (let i = 0; i < pages.length; i++) {
        const page = pages[i];
        const pageObj = page.node;

        if (mode === 'paranoid') {
          this._paranoidRebuildPageInplace(i, pageObj);
        } else {
          this._cleanPageInplace(i, pageObj, mode, context);
        }
      }

      // ── Clean document root ──
      this._cleanRoot(pdfDoc, mode, context, scrubMetadata);

      // ── Metadata Scrubbing (/Info) ──
      if (scrubMetadata) {
        pdfDoc.setTitle('');
        pdfDoc.setAuthor('');
        pdfDoc.setSubject('');
        pdfDoc.setKeywords([]);
        pdfDoc.setProducer('');
        pdfDoc.setCreator('');
        pdfDoc.setCreationDate(new Date(0));
        pdfDoc.setModificationDate(new Date(0));

        const trailer = pdfDoc.context.trailerInfo || pdfDoc.context.trailer;
        if (trailer) {
          if (typeof trailer.get === 'function' && typeof trailer.delete === 'function') {
            const infoId = trailer.get(PDFLib.PDFName.of('Info'));
            if (infoId) {
              trailer.delete(PDFLib.PDFName.of('Info'));
            }
          } else {
            delete trailer['Info'];
            delete trailer.Info;
          }
        }
        this.log('Trailer: Cleared Document Info dictionary (/Info) for metadata scrubbing.');
      }

      // ── Generate PDF bytes ──
      const pdfBytes = await pdfDoc.save();
      const blob = new Blob([pdfBytes], { type: 'application/pdf' });
      this.success = true;
      this.log('PDF sanitization completed successfully.');
      return { success: true, blob };
    } catch (e) {
      this.error(`Sanitization failed: ${e.message}`);
      return { success: false, blob: null };
    }
  }

  _cleanPageInplace(pageIdx, pageObj, mode, context) {
    const keys = Array.from(pageObj.keys());

    for (const key of keys) {
      const norm = this._normalizeName(key.toString());

      // All modes: strip /AA (Additional Actions)
      if (norm === "/AA") {
        pageObj.delete(key);
        this.log(`Page ${pageIdx + 1}: Stripped /AA`);
        continue;
      }

      // Handle page annotations
      if (norm === "/Annots") {
        const annots = this._resolve(pageObj.get(key), context);
        if (annots instanceof PDFLib.PDFArray) {
          this._cleanAnnotationsInplace(pageIdx, annots, mode, context);
        }
      }
    }
  }

  _cleanAnnotationsInplace(pageIdx, annots, mode, context) {
    for (let i = 0; i < annots.size(); i++) {
      const annot = this._resolve(annots.get(i), context);
      if (annot instanceof PDFLib.PDFDict) {
        this._cleanDictObjInplace(`Page ${pageIdx + 1} Annot ${i + 1}`, annot, mode, context);
      }
    }
  }

  _cleanDictObjInplace(location, obj, mode, context) {
    const keys = Array.from(obj.keys());

    for (const key of keys) {
      const norm = this._normalizeName(key.toString());

      // STANDARD: remove high severity keys
      if (PDFSanitizer.HIGH_KEYS.has(norm)) {
        obj.delete(key);
        this.log(`${location}: Stripped key ${norm}`);
        continue;
      }

      // STRICT + PARANOID: also remove medium severity keys
      if (mode === 'strict' || mode === 'paranoid') {
        if (PDFSanitizer.MEDIUM_KEYS.has(norm)) {
          obj.delete(key);
          this.log(`${location}: Stripped key ${norm}`);
          continue;
        }
      }

      // Handle Actions /A sub-dictionary
      if (norm === "/A") {
        const action = this._resolve(obj.get(key), context);
        if (action instanceof PDFLib.PDFDict) {
          const sVal = this._resolve(action.get(PDFLib.PDFName.of('S')), context);
          if (sVal instanceof PDFLib.PDFName) {
            const sType = this._normalizeName(sVal.toString());

            // Standard: remove JS / Launch actions
            if (["/JavaScript", "/JS", "/Launch"].includes(sType)) {
              obj.delete(key);
              this.log(`${location}: Stripped action /S=${sType}`);
              continue;
            }

            // Strict / Paranoid: also remove URI and form submission actions
            if (mode === 'strict' || mode === 'paranoid') {
              if (["/URI", "/SubmitForm", "/ImportData"].includes(sType)) {
                obj.delete(key);
                this.log(`${location}: Stripped action /S=${sType}`);
                continue;
              }
            }
          }
        }
      }

      // Recurse into sub-dictionaries
      const value = this._resolve(obj.get(key), context);
      if (value instanceof PDFLib.PDFDict) {
        this._cleanDictObjInplace(location, value, mode, context);
      }
    }
  }

  _paranoidRebuildPageInplace(pageIdx, pageObj) {
    const keys = Array.from(pageObj.keys());

    for (const key of keys) {
      const norm = this._normalizeName(key.toString());
      if (!PDFSanitizer.PARANOID_SAFE_PAGE_KEYS.has(norm)) {
        pageObj.delete(key);
        this.log(`Page ${pageIdx + 1} (paranoid): Dropped key ${norm}`);
      }
    }
  }

  _cleanRoot(pdfDoc, mode, context, scrubMetadata = false) {
    const catalog = pdfDoc.catalog;
    if (!catalog) return;

    const keys = Array.from(catalog.keys());

    for (const key of keys) {
      const norm = this._normalizeName(key.toString());

      // Skip metadata stream if scrubbing
      if (norm === '/Metadata' && scrubMetadata) {
        catalog.delete(key);
        this.log('Root: Stripped /Metadata stream (Metadata Scrubbing)');
        continue;
      }

      // STANDARD: strip high severity catalog keys
      if (PDFSanitizer.HIGH_KEYS.has(norm)) {
        catalog.delete(key);
        this.log(`Root: Stripped key ${norm}`);
        continue;
      }

      // Clean /Names subtree
      if (norm === "/Names") {
        const names = this._resolve(catalog.get(key), context);
        if (names instanceof PDFLib.PDFDict) {
          this._cleanNamesTree(names, mode, context);
        }
        continue;
      }

      // STRICT + PARANOID: strip medium-severity keys
      if (mode === 'strict' || mode === 'paranoid') {
        if (PDFSanitizer.MEDIUM_KEYS.has(norm)) {
          catalog.delete(key);
          this.log(`Root: Stripped key ${norm}`);
          continue;
        }
      }

      // PARANOID: only keep safe catalog keys
      if (mode === 'paranoid') {
        if (!PDFSanitizer.PARANOID_SAFE_ROOT_KEYS.has(norm)) {
          catalog.delete(key);
          this.log(`Root (paranoid): Dropped key ${norm}`);
        }
      }
    }
  }

  _cleanNamesTree(names, mode, context) {
    const keys = Array.from(names.keys());

    for (const key of keys) {
      const norm = this._normalizeName(key.toString());

      // Always strip JS name tree
      if (norm === "/JavaScript") {
        names.delete(key);
        this.log("Root/Names: Stripped /JavaScript name tree");
        continue;
      }

      // Strict + Paranoid: strip embedded files
      if (mode === 'strict' || mode === 'paranoid') {
        if (norm === "/EmbeddedFiles") {
          names.delete(key);
          this.log("Root/Names: Stripped /EmbeddedFiles");
          continue;
        }
      }
    }
  }
}

window.PDFSanitizer = PDFSanitizer;
